### PR TITLE
Fixes #4918: avoid bogus ellipsis from float noise

### LINKED ISSUE #4918: Total doesn't display the correct value
## What went wrong? 😲

Numbers don't show properly in different cases, depending on what numbers get entered

Screenshot 2024-09-30 at 5 32 36 PM

Screenshot 2024-09-30 at 5 38 12 PM

## Expected behaviour 🤔

They should show normally without the three dots at the end :)

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Go to Dispensary -> Prescriptions
2. Create a new Prescription or edit an existing one
3. Add an item -> under Pack Qty, try and add decimal numbers
4. Save the changes -> see inconsistent number display

## Your environment 🌱

- Open mSupply Version: V2.3.0-RC3
- Legacy mSupply Central Server Version: V7.17.10

- Platform: Tablet (apk)

- Database type: SQLite3

### PR DESCRIPTION
**Pull request by ksawaneh**
Fixes #4918

The ellipsis ("...") marker for >2dp values was triggering for values like `4.40000000000006`, causing totals/currency cells to show `4.40...`.

- Make `NumUtils.hasMoreThanDp/hasMoreThanTwoDp` robust to floating-point rounding noise
- Add unit tests covering the regression

Tests:
- `docker run --rm -v "$PWD":/work -w /work/client node:20-bullseye bash -lc "yarn lint-and-format"`
- `docker run --rm -v "$PWD":/work -w /work/client node:20-bullseye bash -lc "yarn test"`

### CODE REVIEW

**Comment by ksawaneh** [comment]
Added additional unit tests covering very large numbers, very small numbers, and negative numbers for `hasMoreThanTwoDp`/`hasMoreThanDp` in `NumUtils.test.ts` as requested. Rebasing is now up to date with `develop`. Verified with `cd client && yarn test --testPathPattern NumUtils.test.ts` and `cd client && yarn lint-and-format`.

File: client/packages/common/src/utils/numbers/NumUtils.test.ts [hunk 1]
```diff
@@ -20,4 +20,55 @@ describe('NumUtils', () => {
...
+    expect(NumUtils.hasMoreThanDp(4.4, 2)).toBe(false);
+    expect(NumUtils.hasMoreThanDp(4.40000000000006, 2)).toBe(false);
+    expect(NumUtils.hasMoreThanDp(4.41, 2)).toBe(false);
+    expect(NumUtils.hasMoreThanDp(4.411, 2)).toBe(true);
+
+    expect(NumUtils.hasMoreThanDp(0.1234, 3)).toBe(true);
+    expect(NumUtils.hasMoreThanDp(0.123, 3)).toBe(false);
+  });
+
+  it('handles very large numbers', () => {
```
Comment (esmehm) [inline client/packages/common/src/utils/numbers/NumUtils.test.ts]: Thanks for adding these tests! As these changes grow the size of the file, structure is more important - having flat `it(...)` blocks are hard to scan and would be better grouped under one `describe` block per function. 
For example,

```
describe('hasMoreThanTwoDp', () => {
    it('detects values with > 2 decimal places', () => {
      expect(NumUtils.hasMoreThanTwoDp(4.4)).toBe(false);
      expect(NumUtils.hasMoreThanTwoDp(4.40000000000006)).toBe(false);
      expect(NumUtils.hasMoreThanTwoDp(4.401)).toBe(true);
      expect(NumUtils.hasMoreThanTwoDp(0.01)).toBe(false);
      expect(NumUtils.hasMoreThanTwoDp(0.001)).toBe(true);
    });

    it('handles large numbers', () => {
      const large = 123456789.12;
      const largeWithNoise = large + 1e-8;

      expect(NumUtils.hasMoreThanTwoDp(large)).toBe(false);
      expect(NumUtils.hasMoreThanTwoDp(largeWithNoise)).toBe(false);
....
```

File: client/packages/common/src/utils/numbers/NumUtils.test.ts [hunk 2]
```diff
@@ -20,4 +20,55 @@ describe('NumUtils', () => {
...
+    expect(NumUtils.hasMoreThanDp(0.0001, 4)).toBe(false);
+  });
+
+  it('handles negative numbers', () => {
+    expect(NumUtils.hasMoreThanTwoDp(-4.4)).toBe(false);
+    expect(NumUtils.hasMoreThanTwoDp(-4.40000000000006)).toBe(false);
+    expect(NumUtils.hasMoreThanTwoDp(-4.401)).toBe(true);
+
+    expect(NumUtils.hasMoreThanDp(-0.1234, 3)).toBe(true);
+    expect(NumUtils.hasMoreThanDp(-0.123, 3)).toBe(false);
```
Comment (esmehm) [inline client/packages/common/src/utils/numbers/NumUtils.test.ts]: Please also add negative near zero case and test both sides of the boundary for it, eg:
```
expect(NumUtils.hasMoreThanDp(-0.0001, 4)).toBe(false);
expect(NumUtils.hasMoreThanDp(-0.00001, 4)).toBe(true);
```

**Review by esmehm — CHANGES_REQUESTED** [review]
Generally good test coverage! 
However as this test file expands it becomes harder to scan and so please review the structure using describe and it statements to group by function as discussed in the comments. This will also allow you to have a comprehensive test suite for `hasMoreThanDp` and jusrt test that `hasMoreThanTwoDp` is working as a wrapper of the former

**Comment by ksawaneh** [comment]
Thanks! I’ve restructured `NumUtils.test.ts` to group tests by function using `describe(...)`. The edge-case coverage (large/small/negative) now lives under `hasMoreThanDp`, and `hasMoreThanTwoDp` is only tested as a wrapper via `hasMoreThanDp(value, 2)`. Also added the negative near-zero boundary cases and updated the large-number stress test to use `1e15 + 0.12`.
